### PR TITLE
[MRG] Conversion of somato data to BIDS

### PR DESCRIPTION
**Pull request by sappelhoff**
for reference, see this [gitter chat](https://gitter.im/mne-tools/mne-gsoc-2019-BIDS?at=5cf55b1dff3f016baa9b9993)

EDIT: check out the [conversion script](https://github.com/mne-tools/mne-scripts/blob/master/bids_conversion/convert_somato_data.py)

Current structure:

```
MNE-somato-data
├── CHANGES
├── code
│   ├── convert_somato_data.py
│   └── README
├── dataset_description.json
├── derivatives
│   ├── freesurfer
│   │   └── subjects
│   │       ├── 01
│   │       │   ├── bem
│   │       │   ├── label
│   │       │   ├── mri
│   │       │   ├── scripts
│   │       │   ├── stats
│   │       │   ├── surf
│   │       │   ├── tmp
│   │       │   ├── touch
│   │       │   └── trash
│   │       └── fsaverage -> /usr/local/freesurfer/subjects/fsaverage
│   └── sub-01
│       └── sub-01_task-somato-fwd.fif
├── participants.json
├── participants.tsv
├── README
└── sub-01
    ├── anat
    │   ├── sub-01_T1w.json
    │   └── sub-01_T1w.nii.gz
    ├── meg
    │   ├── sub-01_coordsystem.json
    │   ├── sub-01_task-somato_channels.tsv
    │   ├── sub-01_task-somato_events.tsv
    │   ├── sub-01_task-somato_meg.fif
    │   └── sub-01_task-somato_meg.json
    └── sub-01_scans.tsv

19 directories, 16 files

```

### CODE REVIEW

**Comment by sappelhoff** [comment]
It seems like the BIDS naming clashes with the MNE naming convention:

image


This reminds me of: https://github.com/mne-tools/mne-python/issues/4950

**Comment by agramfort** [comment]
yes this has never been fixed in mne-python but it should not warn for
raw bids files

**Comment by sappelhoff** [comment]
> yes this has never been fixed in mne-python but it should not warn for raw bids files

I'll give it a shot

**Comment by jasmainak** [comment]
can you share link to rendered examples?

**Comment by sappelhoff** [comment]
> can you share link to rendered examples?

sure, but they did not run completely due to the runtime warning that the filename for RAW does not follow MNE conventions:

- https://13544-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/time-freq/plot_sensors_time_frequency.html#tut-sensors-time-freq
- https://13544-1301584-gh.circle-artifacts.com/0/dev/auto_examples/inverse/plot_dics_source_power.html#ex-inverse-source-power
- https://13544-1301584-gh.circle-artifacts.com/0/dev/auto_examples/time_frequency/plot_time_frequency_global_field_power.html#ex-time-freq-global-field-power

**Comment by jasmainak** [comment]
ah okay, let us know when that works.

Also I think I prefer to change mne_source_alignment to mne_coregistration

File: examples/inverse/plot_dics_source_power.py [hunk 1]
```diff
@@ -32,9 +36,8 @@
 ###############################################################################
 # Reading the raw data and creating epochs:
 data_path = somato.data_path()
-raw_fname = data_path + '/MEG/somato/sef_raw_sss.fif'
-fname_fwd = data_path + '/MEG/somato/somato-meg-oct-6-fwd.fif'
-subjects_dir = data_path + '/subjects'
+subj = 'sub-01'
+raw_fname = op.join(data_path, subj, 'meg', 'sub-01_task-somato_meg.fif')
```
Comment (jasmainak) [inline examples/inverse/plot_dics_source_power.py]: use variable `subj` from above?

**Comment by jasmainak** [comment]
I would try to be as much close to [BEP003](https://docs.google.com/document/d/17ebopupQxuRwp7U7TFvS6BH03ALJOgGHufxK8ToAvyI/edit?usp=sharing) as possible. So we need to do minimal updates later. For example, you need dataset_description.json:

> Derived dataset and pipeline description
In keeping with the BIDS-Raw standard, a dataset_description.json file MUST be found at the top level of the particular pipeline:

Similarly:

> Each derivative file SHOULD be described by a JSON file provided as a sidecar or higher up in the hierarchy of the derived dataset

**Comment by sappelhoff** [comment]
I agree,

the most current BEP for derivatives is here: https://github.com/bids-standard/bids-specification/pull/207

**Comment by sappelhoff** [comment]
Would these be the correct places to change the naming convention (i.e., accept BIDS endings that are `*_meg.fif`)

https://github.com/mne-tools/mne-python/blob/e1a160c88b8409c8a582e8993cff4806fc0a921e/mne/io/base.py#L1572-L1574

and

https://github.com/mne-tools/mne-python/blob/e1a160c88b8409c8a582e8993cff4806fc0a921e/mne/io/fiff/raw.py#L136-L138


It would also require changing a couple of docstrings and also some docs, such as

- https://github.com/mne-tools/mne-python/blob/master/doc/tutorials/report.rst

**Comment by jasmainak** [comment]
yep! Don't break the report while you do so. I think it expects certain file format in report.parse_folders() :-)

**Comment by sappelhoff** [comment]
@larsoner the new somato data is not found for building the docs ... could this be because we are caching it and the cache is still old?

I changed the md5 hash and the URL though:

https://github.com/mne-tools/mne-python/pull/6414/files#diff-00583e36a75a694e071f9fc054a71c08

**Comment by larsoner** [comment]
The error is:
```
FileNotFoundError: [Errno 2] No such file or directory: '/home/circleci/mne_data/MNE-somato-data/sub-01/meg/sub-01_task-somato_meg.fif'
```
When I download your `tar.gz` and look at the contents I do not see a `MNE-somato-data` folder but rather a `MNE-somato-data-bids` folder. Confirmed this is likely the problem with:
```
$ python -c "import mne; mne.datasets.somato.data_path(force_update=True, verbose=True)"
Downloading archive MNE-somato-data.tar.gz to /home/larsoner/mne_data
...
$ ls ~/mne_data/
...   MNE-somato-data-bids ...
```

**Comment by sappelhoff** [comment]
:man_facepalming: okay, I re-uploaded with proper folder naming. Let's see how this goes.

**Comment by jasmainak** [comment]
@sappelhoff did you run the dataset through the validator? any issues you faced? Is it compliant?

otherwise PR looks clean to me.

**Comment by sappelhoff** [comment]
yep, see:

```Shell
(base) stefanappelhoff@arc-lin-004309:~/Desktop$ bids-validator MNE-somato-data
This dataset appears to be BIDS compatible.

        Summary:                  Available Tasks:        Available Modalities: 
        11 Files, 334.48MB        somato                  T1w                   
        1 - Subject                                       coordsystem           
        1 - Session                                       channels              
                                                          events                
                                                          meg                   


	If you have any questions, please post on https://neurostars.org/tags/bids.
```

I also added `dataset_description.json` files for the derivatives ... they are not completely compliant, but as much as is reasonable given the WIP state of derivatives.

**Comment by jasmainak** [comment]
sounds good. +1 to merge from my end once CIs are happy

**Comment by teonbrooks** [comment]
just curious @sappelhoff, are you adding these derivative files to the BEP 021?

File: examples/inverse/plot_dics_source_power.py [hunk 2]
```diff
@@ -8,19 +8,23 @@
 Compute a Dynamic Imaging of Coherent Sources (DICS) [1]_ filter from
 single-trial activity to estimate source power across a frequency band. This
 example demonstrates how to source localize the event-related synchronization
-(ERS) of beta band activity in the "somato" dataset.
+(ERS) of beta band activity in the
+`somatosensory dataset <https://mne-tools.github.io/stable/manual/datasets_index.html?#somatosensory>`_
```
Comment (agramfort) [inline examples/inverse/plot_dics_source_power.py]: you should not link to absolute html links but link to sphinx anchors.
Also https://mne-tools.github.io is not the main documentation page (it's martinos.org/mne)

File: examples/time_frequency/plot_time_frequency_global_field_power.py [hunk 3]
```diff
@@ -27,6 +27,9 @@
 importantly, the clear-cut comparability of the spectral decomposition (the
 same type of filter is used across all bands).
 
+We will use the
+`somatosensory dataset <https://mne-tools.github.io/stable/manual/datasets_index.html?#somatosensory>`_
```
Comment (agramfort) [inline examples/time_frequency/plot_time_frequency_global_field_power.py]: same remark

**Comment by sappelhoff** [comment]
@teonbrooks not yet - but we just had a BIDS discussion at OHBM and we'll get more focused on the derivatives starting soon :-)

... probably with a survey of what most people need (doing a "data driven 80-20" as @jasmainak put it ;-) )

**Comment by sappelhoff** [comment]
@agramfort with using anchors: do you mean I should do it like this? (see below)

```rst
.. _`somatosensory dataset`: https://martinos.org/mne/stable/manual/datasets_index.html?#somatosensory

=========================================
Compute source power using DICS beamfomer
=========================================

Compute a Dynamic Imaging of Coherent Sources (DICS) [1]_ filter from
single-trial activity to estimate source power across a frequency band. This
example demonstrates how to source localize the event-related synchronization
(ERS) of beta band activity in the `somatosensory dataset`_
```

**Comment by agramfort** [comment]
you need to insert

.. _somato-dataset:

in L124 in datasets_index.rst

and then you'll be able to refer to this section with :ref:`somato-dataset`

>

**Comment by sappelhoff** [comment]
>  you need to insert  .. _somato-dataset:  in L124 in datasets_index.rst  and then you'll be able to refer to this section with :ref:`somato-dataset`
--



to remind myself, this is the rst docs section dealing with that: https://www.sphinx-doc.org/en/1.5/markup/inline.html#cross-referencing-arbitrary-locations

**Comment by agramfort** [comment]
diff looks good. Can you see why CIs fail @sappelhoff ?

**Comment by sappelhoff** [comment]
Yes, the updated data is not yet uploaded. I have to run 
```
mne make_scalp_surfaces -s
mne watershed_bem -s
```

on the new freesurfer derivatives (where subject is `01`, and NOT `sub-01`)

I might need your help later, if I can't make it work :-|

**Comment by jasmainak** [comment]
@sappelhoff why is Circle CI failing here?

**Comment by sappelhoff** [comment]
> @sappelhoff why is Circle CI failing here?

The new data are not yet uploaded, see https://github.com/mne-tools/mne-python/pull/6414#issuecomment-514169581 this will change on Monday when I have access to my work machine again

**Comment by sappelhoff** [comment]
rendered:

1. https://14457-1301584-gh.circle-artifacts.com/0/dev/auto_examples/inverse/plot_dics_source_power.html
1. https://14457-1301584-gh.circle-artifacts.com/0/dev/auto_examples/time_frequency/plot_time_frequency_global_field_power.html
1. https://14457-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/time-freq/plot_sensors_time_frequency.html

File: examples/inverse/plot_dics_source_power.py [hunk 4]
```diff
@@ -79,4 +86,4 @@
 stc = beta_source_power / baseline_source_power
 message = 'DICS source power in the 12-30 Hz frequency band'
 brain = stc.plot(hemi='both', views='par', subjects_dir=subjects_dir,
-                 time_label=message)
+                 subject=subj, time_label=message)
```
Comment (agramfort) [inline examples/inverse/plot_dics_source_power.py]: this tells me that subj should be renamed subject everywhere

**Comment by jasmainak** [comment]
Thanks @sappelhoff !

File: examples/inverse/plot_dics_source_power.py [hunk 5]
```diff
@@ -32,9 +34,10 @@
 ###############################################################################
 # Reading the raw data and creating epochs:
 data_path = somato.data_path()
-raw_fname = data_path + '/MEG/somato/sef_raw_sss.fif'
-fname_fwd = data_path + '/MEG/somato/somato-meg-oct-6-fwd.fif'
-subjects_dir = data_path + '/subjects'
+subject = '01'
+task = 'somato'
+raw_fname = op.join(data_path, 'sub-{}'.format(subject), 'meg',
+                    'sub-{}_task-{}_meg.fif'.format(subject, task))
```
Comment (massich) [inline examples/inverse/plot_dics_source_power.py]: I know I'm late on the review cycle and this is merged, but do you guys find worth to create a helper function that allows composing the name?

```py
def fname_helper(subject, task='somato', data_type_prefix=''):
    return op.join(
        'sub-{0:02d}'.format(subject),
        data_type_prefix,
        'sub-{0:02d}_task-{}_meg.fif'.format(subject, task)
    )
```

```py
from mne.datasets.somato import data_path, fname_helper

raw_fname = op.join(data_path(), fname_helper(subject=1, data_type_prefix='meg'))
fname_fwd = op.join(data_path, 'derivatives', fname_helper(subject=1, data_type_prefix='meg'))
```

Comment (massich) [reply examples/inverse/plot_dics_source_power.py]: ups !! the file names are not the same. 

one is `_meg.fif` the other one is `_fwd.fif`

**Comment by massich** [comment]
Also reading the artifacts I found that we are using a private function in a tutorial:

https://github.com/mne-tools/mne-python/blob/cc836c2635b7a720ec6c60074a1997fe52d7ca2f/examples/time_frequency/plot_time_frequency_global_field_power.py#L56

Shall we make the function public? wrap wherever is exposed and make this one public?

**Comment by agramfort** [comment]
>
> Also reading the artifacts I found that we are using a private function in
> a tutorial:
>
>
> https://github.com/mne-tools/mne-python/blob/cc836c2635b7a720ec6c60074a1997fe52d7ca2f/examples/time_frequency/plot_time_frequency_global_field_power.py#L56
>
> Shall we make the function public? wrap wherever is exposed and make this
> one public?
>

nice catch ! this is evil. @sappelhoff can you fix? or anyone?

**Comment by sappelhoff** [comment]
Can @massich elaborate a bit on his intention for the fix?

File: examples/inverse/plot_dics_source_power.py (hunk 5 above)
Comment (jasmainak) [reply examples/inverse/plot_dics_source_power.py]: The derivatives is not set in stone, so I would not make a new function. There is already one in mne-bids for the existing BIDS spec: https://mne.tools/mne-bids/generated/mne_bids.make_bids_basename.html#mne_bids.make_bids_basename.

**Comment by jasmainak** [comment]
you're not supposed to use private functions in examples ...

**Comment by sappelhoff** [comment]
Yes, I learned that repeatedly through you and Alex :-D 

But I was wondering what @massich  meant with

> wrap wherever is exposed and make this one public?

**Comment by jasmainak** [comment]
well if you wrap another function around it, you can potentially expose fewer parameters publicly while still keeping the flexibility of the private function. Not sure if that's necessary though.

**Comment by larsoner** [comment]
@sappelhoff looks like the `-5120-bem.fif` and `-5120-bem-sol.fif` files are missing from the `bem` directory. Was this intentional or just an oversight? If the latter, I could add them but do I need to do anything at the BIDS end? FWIW this can be used to create them:
```Python
# in ~/mne_data/MNE-somato-data/derivatives/freesurfer/subjects
model = mne.make_bem_model('01', conductivity=(0.3,), subjects_dir='.', verbose=True)
mne.write_bem_surfaces('01/bem/01-5120-bem.fif', model)
bem = mne.make_bem_solution(model, verbose=True)
mne.write_bem_solution('01/bem/01-5120-bem-sol.fif', bem)

**Comment by britta-wstnr** [comment]
Can confirm -- tried to run some older code after updating `somato`. The BEM solution seems to be missing.

**Comment by sappelhoff** [comment]
> Was this intentional or just an oversight?

sounds like an oversight to me because I have no clue what these files do.

This is the conversion script: https://github.com/mne-tools/mne-scripts/blob/master/bids_conversion/convert_somato_data.py

As long as the files are being put into the `derivatives` directory, it does not hurt BIDS. But the best would probably be to fix the conversion script.

**Comment by larsoner** [comment]
Okay, will push a commit there and update the dataset on OSF

**Comment by sappelhoff** [comment]
> Okay, will push a commit there and update the dataset on OSF

thanks a lot @larsoner and sorry to cause problems.

**Pull request merged**